docs/site/LB3-vs-LB4-request-response-cycle.md

```diff
@@ -0,0 +1,448 @@
+- `final`
+- `final:after`
+
+Any middleware higher up in the chain may terminate the request by sending a
+response back to the client.
+
+![LoopBack 3 request/response components](./imgs/lb3-req-res.png)
+
+The REST API middleware is added in the `routes` phase and error handlers in the
+`final` and `final:after` phases.
+
+The REST API middleware is responsible for creating REST API endpoints out of
+the models in the app. It then uses the configured datasource for connecting
+and querying the undelying database for the corresponding REST requests.
+
+In LoopBack 4, the [sequence](https://loopback.io/doc/en/lb4/Sequence.html) is
+the gatekeeper of all requests to the app. Every request to the app must pass
+through the sequence. It identifies the responsible handler for the requested
+endpoint and passes on the request to the handler.
+
+![LoopBack 4 request/response components](./imgs/lb4-req-res.png)
+
+Unlike LoopBack 3, models in LoopBack 4 do not create REST endpoints. Use the
+`lb4 controller` command to quickly generate the REST endpoints for a model.
+
+For more details, refer to the LoopBack 4
+[request/response cycle](https://loopback.io/doc/en/lb4/Request-response-cycle.html)
+doc.
+
+#### Access to the request/response object
+
+Since LoopBack 3 uses the Express middleware pattern, the request and response
+objects can always be accessed in all the middleware functions.
+
+In LoopBack 4, the request and response objects can be accessed in
+routers loaded using the
+[app.mountExpressRouter() ](https://loopback.io/doc/en/lb4/apidocs.rest.restapplication.mountexpressrouter.html)
+method, using the familiar Express middleware signature.
+
+Controllers, services, and repositories are LoopBack 4 components that
+participate in the request/response cycle. The request and response objects can
+be made available to them via
+[dependency injection](https://loopback.io/doc/en/lb4/Dependency-injection.html)
+.
+
+Example of accesssing the request and response object in a Controller:
+
+```ts
+import {inject} from '@loopback/context';
+import {Request, Response, RestBindings, get} from '@loopback/rest';
+
+export class ExampleController {
+  constructor(
+    @inject(RestBindings.Http.REQUEST) private req: Request,
+    @inject(RestBindings.Http.RESPONSE) private res: Response
+    ) {}
+
+  @get('/headers')
+  headers() {
+    // Sends back the request headers
+    this.res.send(this.req.headers);
+  }
+}
+```
+
+Similarly, the request and response objects can be injected into services
+and respositories along with other objects from the
+[context](https://loopback.io/doc/en/lb4/Context.html).
+
+{% include tip.html content="It may be tempting to use an Express router
+(because of familiarity) instead of a controller to add custom endpoints to the
+app, but bear it in mind that controllers are way more powerful and capable than
+Express routers because of their support for dependency injection and access to
+the application context." %}
```

Comment:
IMO, the most important reason for preferring Controllers over Express routes is OpenAPI spec:

1. With Controllers, the framework parses and validates the incoming request parameters based on the provided OpenAPI spec.
2. Routes contributed by controllers are included in the auto-generated OpenAPI spec document.

It's possible to provide OpenAPI spec documentation when mounting an Express router, however the routes must handle parameter parsing & validation themselves.

Please include this information in the tip.